packages/vira/src/Vira/Page/JobPage.hs

```diff
@@ -153,35 +179,94 @@ triggerNewBuild repoName branchName = do
     App.update $ St.JobUpdateStatusA job.jobId St.JobRunning
     log Info $ "Started task " <> show job.jobId
 
--- | Get all build stages
-getStages :: St.Repo -> St.Branch -> Maybe CachixSettings -> Maybe AtticSettings -> NonEmpty CreateProcess
-getStages repo branch mCachix mAttic = do
-  stageCreateProjectDir
-    :| one stagesClone
-    <> maybe [] (one . stageAtticLogin) mAttic
-    <> [stageBuild]
-    <> (maybe mempty (one . stageCachixPush) mCachix <> maybe mempty (one . stageAtticPush) mAttic)
+-- TODO: Move to appropriate module
+newtype RepoSettings = RepoSettings
+  { stages :: [Stage]
+  -- ^ All stages in a `Task`
+  }
+  deriving stock (Show)
+
+-- | `Action` with conditions
+data Stage = Stage
+  { conditions :: [ActionCondition]
+  , action :: Action
+  }
+  deriving stock (Show)
+
+-- | User-configurable action in a `Task`
+data Action
+  = Clone St.Repo St.Branch
+  | AtticLogin AtticSettings
+  | Build BuildSettings
+  | AtticPush AtticSettings
+  | CachixPush CachixSettings
+  deriving stock (Show)
+
+-- | Settings for the build `Action`
+newtype BuildSettings = BuildSettings
+  { extraArgs :: [Text]
+  -- ^ extra CLI arguments to the build command
+  }
+  deriving stock (Show)
+
+-- | Condition for when to run an `Action`
+newtype ActionCondition
+  = -- | Whether the branch name of the current checkout matches the given pattern
+    BranchMatches Text
+  deriving stock (Show)
+
+-- | Execution order of an `Action`
+actionOrder :: Action -> Int
+actionOrder = \case
+  Clone _ _ -> 0
+  AtticLogin _ -> 1
+  Build _ -> 2
+  AtticPush _ -> 3
+  CachixPush _ -> 4
+
+-- | Return final actions to run in a `Task`
+getActions :: BranchName -> [Stage] -> [Action]
+getActions branchName =
+  uniqueSortOn actionOrder
+    . map action
+    . filter (all match . conditions)
+  where
+    match :: ActionCondition -> Bool
+    match (BranchMatches p) = matchWildcard (toString p) (toString branchName.unBranchName)
```

Comment:
Worth moving this to top-level associating close by `ActionCondition` even if only for documentation/readability purposes (a function in `where` clause uses parent scope's variables which is not always obvious).